android/src/main/java/com/rudderstack/android/internal/plugins/SessionPlugin.kt

```diff
@@ -0,0 +1,91 @@
+ * permissions and limitations under the License.
+ */
+
+package com.rudderstack.android.internal.plugins
+
+import com.rudderstack.android.ConfigurationAndroid
+import com.rudderstack.android.internal.extensions.withSessionId
+import com.rudderstack.android.internal.extensions.withSessionStart
+import com.rudderstack.android.internal.states.UserSessionState
+import com.rudderstack.android.utilities.defaultLastActiveTimestamp
+import com.rudderstack.android.utilities.resetSession
+import com.rudderstack.android.utilities.startSessionIfNeeded
+import com.rudderstack.android.utilities.updateSessionEnd
+import com.rudderstack.core.Analytics
+import com.rudderstack.core.Configuration
+import com.rudderstack.core.Plugin
+import com.rudderstack.core.internal.states.ConfigurationsState
+import com.rudderstack.models.Message
+import com.rudderstack.models.android.UserSession
+
+internal class SessionPlugin : Plugin {
+
+    private var _analytics: Analytics? = null
+    override fun setup(analytics: Analytics) {
+        super.setup(analytics)
+        _analytics = analytics
+        ConfigurationsState.subscribe(::configurationSubscriber)
+    }
+
+    private fun configurationSubscriber(
+        newConfiguration: Configuration?,
+        oldConfiguration: Configuration?
+    ) {
+        if (newConfiguration !is ConfigurationAndroid || oldConfiguration !is ConfigurationAndroid) return
+        if (newConfiguration.trackLifecycleEvents != oldConfiguration.trackLifecycleEvents
+            || newConfiguration.trackAutoSession != oldConfiguration.trackAutoSession) {
+            if(newConfiguration.trackAutoSession && newConfiguration.trackLifecycleEvents) _analytics?.startSessionIfNeeded()
+            else _analytics?.updateSessionEnd()
+        }
+    }
+
+    override fun intercept(chain: Plugin.Chain): Message {
+        //apply session id and session start
+        // update last active timestamp
+        // if difference between two events is more than session timeout, refresh session
+        val message = chain.message()
+        _analytics?.startSessionIfNeeded()
+        val newMsg = UserSessionState.value?.takeIf { it.isActive }?.let {
+            updateWithSession(message, it)
+        } ?: message
+
+        return chain.proceed(newMsg)
+    }
+
+    private fun updateWithSession(
+        message: Message, it: UserSession
+    ): Message {
+        val context = message.context
+        var newContext = context?.withSessionId(it.sessionId.toString())
+        if (it.sessionStart) newContext = newContext?.withSessionStart(true)
+
+        updateSessionState(it)
+        return message.copy(newContext)
+    }
+
+    private fun updateSessionState(it: UserSession) {
+        val updatedSession = it.copy(
+            lastActiveTimestamp = defaultLastActiveTimestamp, sessionStart = false
+        )
+        UserSessionState.update(updatedSession)
+    }
+
+
+
+    override fun reset() {
+        super.reset()
```

Comment:
☝️ Here (and in other places as well) we are calling an empty interface method. This has been done to avoid several empty callbacks from the Plugins that don't really implement all the interface members. Ηowever it seems a bit weird. 🤔